src/libmongoc/src/mongoc/mongoc-uri.c

```diff
@@ -1423,23 +1423,28 @@ mongoc_uri_parse (mongoc_uri_t *uri, const char *str, bson_error_t *error)
       goto error;
    }
 
-   before_slash = scan_to_unichar (str, '/', "", &tmp);
-   if (!before_slash) {
-      before_slash = bson_strdup (str);
-      str += strlen (before_slash);
+   // Check for delimiting slash (not required)
+   before_options = scan_to_unichar (str, '/', "", &tmp);
+   if (!before_options) {
+      before_options = scan_to_unichar (str, '?', "", &tmp);
```

Comment:
This change may prevent including `?` in the userinfo. Example:

```c
bson_error_t error;
mongoc_uri_t *uri = mongoc_uri_new_with_error ("mongodb://user:pa?s@localhost", &error);
ASSERT_OR_PRINT (uri, error); // Passes before PR. With PR, fails with "Invalid host string in URI".
ASSERT_CMPSTR (mongoc_uri_get_password (uri), "pa?s");
mongoc_uri_destroy (uri);
```

The [Connection String spec](https://github.com/mongodb/specifications/blob/e9f02f328a93d6a0321694cc30d031079c555e4a/source/connection-string/connection-string-spec.md#userinfo-optional) notes:

> drivers SHOULD require connection string user information to follow the "userinfo" encoding rules of RFC 3986 and SHOULD throw an exception if disallowed characters are detected. However, for backwards-compatibility reasons, drivers MAY allow reserved characters other than "@" and ":" to be present in user information without percent-encoding.

RFC 3968 prohibits a `?` in the userinfo, but the C driver permits it. I expect `?` is used in passwords, and given the [difficulties encountered](https://jira.mongodb.org/browse/CDRIVER-4119?focusedCommentId=5518486&focusedId=5518486&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-5518486) with non-ASCII passwords, it would not surprise me if there is real-world use passing literal `?` in the password.

I suggest adding this test case, and changing to preserve accepting `?` in userinfo. Consider trying to find the `?` after a possible `@`.